src/core/global.c

```diff
@@ -570,27 +563,46 @@ int nn_socket (int domain, int protocol)
 int nn_close (int s)
 {
     int rc;
+    struct nn_sock *sock;
 
-    NN_BASIC_CHECKS;
-
-    /*  TODO: nn_sock_term can take a long time to accomplish. It should not be
-        performed under global critical section. */
     nn_glock_lock ();
+    rc = nn_global_hold_socket_locked (&sock, s);
```

Comment:
Describing to someone who would have zero previous knowledge; why is it necessary to "hold" a socket on line 569 when a "global lock" was acquired on line 568?